cpmpy/solvers/pumpkin.py

```diff
@@ -0,0 +1,531 @@
+from ..transformations.decompose_global import decompose_in_tree
+from ..transformations.flatten_model import flatten_constraint
+from ..transformations.comparison import only_numexpr_equality
+from ..transformations.reification import reify_rewrite, only_bv_reifies, only_implies
+from ..transformations.safening import no_partial_functions
+
+import time
+
+"""
+    Interface to Pumpkin's API
+
+    Pumpkin is a combinatorial optimisation solver developed by the ConSol Lab at TU Delft. 
+    It is based on the (lazy clause generation) constraint programming paradigm.
+    (see https://github.com/consol-lab/pumpkin)
+
+    Always use :func:`cp.SolverLookup.get("pumpkin") <cpmpy.solvers.utils.SolverLookup.get>` to instantiate the solver object.
+
+    ===============
+    Installation
+    ===============
+
+    The `pumpkin_py` python package is currently not available on PyPI.
+    It can be installed from source using the following steps:
+     1. clone the repository from github: https://github.com/consol-lab/pumpkin
+     2. install the "maturin" package to build the python bindings: $ pip install maturin
+     3. build and install the package: $ maturin develop
+
+    ===============
+    List of classes
+    ===============
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_pumpkin
+
+    ==============
+    Module details
+    ==============
+"""
+
+class CPM_pumpkin(SolverInterface):
+    """
+    Interface to Pumpkin's API
+
+    Creates the following attributes (see parent constructor for more):
+
+    - ``pum_solver``: the pumpkin.Model() object
+    Creates the following attributes (see parent constructor for more):
+    - tpl_model: object, Pumpkin's model object
+    """
+
+    @staticmethod
+    def supported():
+        # try to import the package
+        try:
+            import pumpkin_py as gp
+            return True
+        except ImportError:
+            return False
+
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: Model(), a CPMpy Model() (optional)
+        - subsolver: None, not used
+        """
+        if not self.supported():
+            raise Exception("CPM_Pumpkin: Install the python package 'pumpkin_py'")
+
+        from pumpkin_py import Model
+
+        assert subsolver is None 
+
+        # initialise the native solver object
+        self.pum_solver = Model() 
+
+        # for objective
+        self._objective = None
+        self.objective_is_min = True
+
+        # initialise everything else and post the constraints/objective
+        super().__init__(name="Pumpkin", cpm_model=cpm_model)
+
+    @property
+    def native_model(self):
+        """
+            Returns the solver's underlying native model (for direct solver access).
+        """
+        return self.pum_solver
+
+
+    def solve(self, time_limit=None, proof=None, assumptions=None):
```

Comment:
so in GCS we already made some decisions for the 'solve' api to proof logging...
there we decided an attribute 'prove' if it should store its proof, and 'verify' if it should be verified... then we also have proof_name and proof_location; I'm not sure these are the best choices, e,g. 'proof=True, proof_file=...' would also make sense, but I do think we should have both solvers have equally named attributes

Reply:
Hmm yeah having both proof-name and proof-location seems a bit redundant. But we shouldn't change the gcs interface in this PR...
So I'll now mimic the gcs interface, and then we change later on if we want.

(Related: probably we can also enable prooflogging in pysat and exact?)